### ISSUE TITLE
kube-proxy-replacement breaks nfs reconnect

### ISSUE THREAD

**Issue by Rid** [cilium/cilium#21541]
### Is there an existing issue for this?

- [X] I have searched the existing issues

Update: This would apply to any in-kernel client, as when the connection is initiated from the kernel a different code-path is used when updating the service IP.

### What happened?

We setup cilium using:
`cilium install --kube-proxy-replacement probe --helm-set k8sServiceHost=1.2.3.4,k8sServicePort=6443,global.etcd.enabled=true,global.etcd.managed=true,operator.replicas=3`

We are using the NFS operator from here:
https://github.com/openebs/dynamic-nfs-provisioner

We provision a pod attached to the NFS mounted volume, everything works successfully.

We then fake an outage on the node hosting the NFS server, the NFS pod is then rescheduled to another server.

When using `kube-proxy-replacement` the mount within the pod consuming the NFS server hangs indefinitely with:

```
[147766.740926] NFSD: Unable to initialize client recovery tracking! (-22)
[147766.740929] NFSD: starting 90-second grace period (net f000174e)
[147792.384961] nfs: server 10.76.145.123 not responding, timed out
[147819.009203] nfs: server 10.76.145.123 not responding, still trying
[147972.610076] nfs: server 10.76.145.123 not responding, timed out
[147977.733893] nfs: server 10.76.145.123 not responding, timed out
[147980.801919] nfs: server 10.76.145.123 not responding, still trying
[148152.834725] nfs: server 10.76.145.123 not responding, timed out
[148157.954807] nfs: server 10.76.145.123 not responding, timed out
[148163.075008] nfs: server 10.76.145.123 not responding, timed out
```

`10.76.145.123` is the cluster IP of the NFS service

```
service/nfs-pvc-729ceb2e-0959-4d81-bc96-94fd37a73061   ClusterIP   10.76.145.123   <none>        2049/TCP,111/TCP   42m   openebs.io/nfs-server=nfs-pvc-729ceb2e-0959-4d81-bc96-94fd37a73061
```

When not using `kube-proxy-replacement` the mount reconnects automatically and everything works again.

### Cilium Version

cilium-cli: v0.12.4 compiled with go1.19.1 on linux/amd64
cilium image (default): v1.12.2
cilium image (stable): v1.12.2
cilium image (running): v1.12.2

### Kernel Version

Ubuntu 22.04

Linux dedi2-cplane1.23-106-58-144.lon-01.uk.appsolo.com 5.15.0-48-generic #54-Ubuntu SMP Fri Aug 26 13:26:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.25.2

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Rid** [comment]
Here's the relevant entry from the service list:

```
258   10.76.145.123:111       ClusterIP      1 => 10.0.3.167:111 (active)
259   10.76.145.123:2049      ClusterIP      1 => 10.0.3.167:2049 (active)
```

It has the correct entry for the rescheduled pod

```
pod/nfs-pvc-729ceb2e-0959-4d81-bc96-94fd37a73061-78d48ffc59-6prlp   1/1     Running       0          34m    10.0.3.167   dedi1-node1.23-106-58-141.lon-01.uk.appsolo.com   <none>           <none>
```

`pwru --filter-dst-ip=10.76.145.123 --output-tuple` shows no traffic also `pwru --filter-dst-ip=10.0.3.167 --output-tuple` shows no traffic.

`conntrack -L | grep 10.76.145.123` shows no entries, `conntrack -E | grep 10.76.145.123` shows no events.

What other logs can I provide?

**Comment by Rid** [comment]
If another mount is attempted from the same node to the service IP, it will try to use the old unavailable pod IP.

However if another mount is attempted from the same node to the server IP, but using a different version of NFS, it will use the new pod IP.

This makes me believe that there is something perhaps in the NFS kernel client which is using different connection routing than eBPF?

**Comment by brb** [comment]
@Rid Thanks for the detailed issue. I think you are hitting https://github.com/cilium/cilium/issues/12859 (cc @aditighag). TL;DR the sock has the old service backend IP. You could work around by either switching NFS to use TCP or by disabling the socket LB for pods (`socketLB.hostNamespaceOnly=true`).

**Comment by Rid** [comment]
@brb Thanks for taking a look, as you can see from the logs above the NFS mounts are already TCP.

```
10.69.146.63:/ on /var/lib/kubelet/pods/fb6a9039-589d-48b8-8b42-215706216ecb/volumes/kubernetes.io~nfs/pvc-714d603f-a2bc-4f61-aab3-5e346c3bb4da type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.69.146.63,local_lock=none,addr=10.69.146.63)
10.68.201.199:/ on /var/lib/kubelet/pods/54c775ea-a909-4811-b676-4f504b343d5c/volumes/kubernetes.io~nfs/pvc-703de151-7b89-4404-ab26-d11b1540a39a type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.68.201.199,local_lock=none,addr=10.68.201.199)
10.78.215.164:/ on /var/lib/kubelet/pods/9b3d5e2d-4544-4976-9b0d-28d737b7bd55/volumes/kubernetes.io~nfs/pvc-748d9915-84dc-4850-bb7b-0532b86d5944 type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=23.106.58.142,local_lock=none,addr=10.78.215.164)
```

I will try with `socketLB.hostNamespaceOnly=true` and report back.

**Comment by Rid** [comment]
@brb Unfortunately even with `socketLB.hostNamespaceOnly=true` we're still having the same problem, so perhaps it's caused by a different issue?

You can reproduce it with the following:

```bash
#!/bin/bash
#
# Edit <backend-storageclass-name> to match your storageclass name, the backend storage
# should be accessible from both nodes, we use https://github.com/piraeusdatastore/piraeus for this purpose
#
# Edit <node1> and <node2> to match your node names

kubectl apply -f https://openebs.github.io/charts/nfs-operator.yaml

kubectl cordon <node2>

kubectl apply -f - <<EOF
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: new-storageclass
  namespace: default
  annotations:
    openebs.io/cas-type: nfsrwx
    cas.openebs.io/config: |
      - name: NFSServerType
        value: "kernel"
      - name: BackendStorageClass
        value: "<backend-storageclass-name>"
allowedTopologies:
- matchLabelExpressions:
  - key: kubernetes.io/hostname
    values:
    - <node1>
    - <node2>
provisioner: openebs.io/nfsrwx
reclaimPolicy: Delete
---
apiVersion: v1
kind: Pod
metadata:
  name: busybox
  namespace: default
spec:
  containers:
  - image: busybox
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
    name: busybox
    volumeMounts:
    - mountPath: /nfs
      name: nfs
  restartPolicy: Always
  volumes:
  - name: nfs
    persistentVolumeClaim:
      claimName: nfs-pvc
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: nfs-pvc
  namespace: default
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 10Mi
  storageClassName: new-storageclass
  volumeMode: Filesystem
EOF

kubectl uncordon <node2>
kubectl cordon <node1>

NFS_POD=$(kubectl -n openebs get all | grep 'pod/nfs-pvc' | awk '{print $1}')

kubectl -n openebs delete ${NFS_POD}

# Wait for the pod to be rescheduled

kubectl uncordon <node1>

kubectl -n default exec -it busybox -- ls /nfs

# This will hang indefinitely now, dmesg on the node will show the timeout:
# nfs: server <NFS SVC IP> not responding, timed out
```

**Comment by Rid** [comment]
ping @brb @aditighag would you be able to run the reproducer and let me know if it's reproducible?

**Comment by brb** [comment]
> Can you confirm that the NFS provisioner is running with

```
> kubectl -n openebs get all -o wide
NAME                                                               READY   STATUS    RESTARTS   AGE   IP             NODE           NOMINATED NODE   READINESS GATES
pod/nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f-f6d7b9f76-hvmfz   1/1     Running   0          40m   10.244.2.247   kind-worker2   <none>           <none>
pod/openebs-nfs-provisioner-66674b9b-9cf5w                         1/1     Running   0          40m   10.244.1.163   kind-worker    <none>           <none>

NAME                                                   TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)            AGE   SELECTOR
service/nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f   ClusterIP   10.96.125.45   <none>        2049/TCP,111/TCP   40m   openebs.io/nfs-server=nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f

NAME                                                           READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS                IMAGES                            SELECTOR
deployment.apps/nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f   1/1     1            1           40m   nfs-server                openebs/nfs-server-alpine:0.9.0   openebs.io/nfs-server=nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f
deployment.apps/openebs-nfs-provisioner                        1/1     1            1           40m   openebs-provisioner-nfs   openebs/provisioner-nfs:0.9.0     name=openebs-nfs-provisioner,openebs.io/component-name=openebs-nfs-provisioner

NAME                                                                     DESIRED   CURRENT   READY   AGE   CONTAINERS                IMAGES                            SELECTOR
replicaset.apps/nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f-f6d7b9f76   1         1         1       40m   nfs-server                openebs/nfs-server-alpine:0.9.0   openebs.io/nfs-server=nfs-pvc-74970974-8d5d-419a-a416-ddc2a7828f4f,pod-template-hash=f6d7b9f76
replicaset.apps/openebs-nfs-provisioner-66674b9b                         1         1         1       40m   openebs-provisioner-nfs   openebs/provisioner-nfs:0.9.0     name=openebs-nfs-provisioner,openebs.io/component-name=openebs-nfs-provisioner,pod-template-hash=66674b9b
```

I ran the following:
```
cd cilium
./contrib/scripts/kind.sh '' 2 '' '' 'none'
cd ../cilium-cli
./cilium install --wait --rollback=false --config monitor-aggregation=none --helm-set-string=kubeProxyReplacement=strict --helm-set-string=tunnel=vxlan --version=v1.12.4
```
And then your `nfs.sh` with the following nodes:
```
set +ex

NODE1=kind-worker
NODE2=kind-worker2
```

**Comment by Rid** [comment]
@brb that's strange, kind-worker2 should have been cordoned so the nfs-pvc pod shouldn't have been scheduled on that node, perhaps we should have waited until the resources were ready, could you try the following?

```sh
#!/bin/bash
#
# Edit ${NODE1} and ${NODE2} to match your node names
#

NODE1=node1-name
NODE2-node2-name

kubectl apply -f https://openebs.github.io/charts/nfs-operator.yaml

kubectl cordon ${NODE2}

kubectl apply -f - <<EOF
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: my-local-storage
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: my-local-pv
spec:
  capacity:
    storage: 10Mi
  accessModes:
  - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: my-local-storage
  local:
    path: /tmp
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/hostname
          operator: In
          values:
          - ${NODE1}
          - ${NODE2}
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: new-storageclass
  namespace: default
  annotations:
    openebs.io/cas-type: nfsrwx
    cas.openebs.io/config: |
      - name: NFSServerType
        value: "kernel"
      - name: BackendStorageClass
        value: "my-local-storage"
allowedTopologies:
- matchLabelExpressions:
  - key: kubernetes.io/hostname
    values:
    - ${NODE1}
    - ${NODE2}
provisioner: openebs.io/nfsrwx
reclaimPolicy: Delete
---
apiVersion: v1
kind: Pod
metadata:
  name: busybox
  namespace: default
spec:
  containers:
  - image: busybox
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
    name: busybox
    volumeMounts:
    - mountPath: /nfs
      name: nfs
  restartPolicy: Always
  volumes:
  - name: nfs
    persistentVolumeClaim:
      claimName: nfs-pvc
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: nfs-pvc
  namespace: default
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 10Mi
  storageClassName: new-storageclass
  volumeMode: Filesystem
EOF

echo "Waiting for pod to be created"
until [ "$(kubectl get all | grep 'pod/busybox' | awk '{print $3}')" = "Running" ]; do sleep 1; done

kubectl uncordon ${NODE2}
kubectl cordon ${NODE1}

NFS_POD=$(kubectl -n openebs get all | grep 'pod/nfs-pvc' | awk '{print $1}')

kubectl -n openebs delete ${NFS_POD}

echo "Waiting for pod to be rescheduled"
until [ "$(kubectl -n openebs get all | grep 'pod/nfs-pvc' | awk '{print $3}')" = "Running" ]; do sleep 1; done

kubectl uncordon ${NODE1}

kubectl -n default exec -it busybox -- ls /nfs

# This will hang indefinitely now, dmesg on the node will show the timeout:
# nfs: server <NFS SVC IP> not responding, timed out
```

**Comment by anthr76** [comment]
I appear to also be facing this issue when hosting a nfs server with Service type LoadBalancer.

**Comment by brb** [comment]
> Is 10.96.185.162 reachable from kind-worker?

Noup. But this gives me a good hint what might be happening. I will comment in a bit.

**Comment by brb** [comment]
@Rid When you hit the issue, was kube-proxy running?

**Comment by Rid** [comment]
> @Rid When you hit the issue, was kube-proxy running?

No it was removed.

**Comment by brb** [comment]
Can you provide the Cilium sysdump?

**Comment by brb** [comment]
@Rid bump

**Comment by Rid** [comment]
> Can you provide the Cilium sysdump?

I'm just setting up a cluster on digital ocean to test the reproducer outside of our setup. I will send you a message on Slack to provide you access.

**Comment by dmeijboom** [comment]
Not sure if related but I'm running Longhorn & Cilium (also with kube proxy replacement) and RWX volumes don't seem to work. Note that Longhorn uses NFS for this, getting similar logs:

```
[20760.312153] nfs: server 10.96.13.86 not responding, timed out
[20765.435864] nfs: server 10.96.13.86 not responding, timed out
[20770.552228] nfs: server 10.96.13.86 not responding, timed out
[20807.420819] nfs: server 10.96.13.86 not responding, timed out
[20825.209258] nfs: server 10.96.13.86 not responding, still trying
```

**Comment by dhedberg** [comment]
We're having the same issue; existing mounts hang when the pod behind the service is replaced with a new pod with a different IP.

I've been looking into it with bpftrace, and I think the issue is that the sunrpc in xprtsock.c calls kernel_connect with a ```struct sockaddr``` that is actually a pointer into a long-living ```struct rpc_xprt```. Some output below from my tracing:

```
[ The NFS-share is mounted. The service address is 10.102.48.51, the pod is 192.168.4.3. ]

sunrpc:
xs_connect: (xprt: 0xffff93f93abd6800) 10.102.48.51
xs_tcp_finish_connecting: (xprt: 0xffff93f93abd6800) 10.102.48.51

kernel:
kernel_connect: 10.102.48.51
inet_stream_connect: (addr: 0xffff93f93abd6820) 10.102.48.51
tcp_v4_pre_connect: (addr: 0xffff93f93abd6820) 10.102.48.51
[ address presumably rewritten by cilium in cil_sock4_connect() ]
tcp_v4_connect: (addr: 0xffff93f93abd6820) 192.168.4.3

[ tcpkill and a result reconnect; note that xs_connect now uses the pod address directly ]

sunrpc:
xs_connect: (xprt: 0xffff93f93abd6800) 192.168.4.3
xs_tcp_finish_connecting: (xprt· 0xffff93f93abd6800) 192.168.4.3

kernel:
kernel_connect: 192.168.4.3
inet_stream_connect: (addr: 0xffff93f93abd6820) 192.168.4.3
tcp_v4_pre_connect: (addr: 0xffff93f93abd6820) 192.168.4.3
tcp_v4_connect: (addr: 0xffff93f93abd6820) 192.168.4.3
```

I don't know if this is a kernel bug (does sunrpc need to make a copy of the address?) or a case of cilium doing something that's not going to work.

**Comment by Rid** [comment]
@dhedberg That's interesting, do you have any idea where `xprt` is being updated to the pod IP.

When you're doing the tcpkill, has the pod already been replaced by a new one?

I have sent @brb the details to a test cluster which reproduces the issue, we're currently waiting for him to debug.

**Comment by dhedberg** [comment]
@Rid Yes, the issue is that there's only one copy of the address involved here. So any changes to this address will actually change the address in rpc_xprt.

My understanding is:

* 1. The final step in tcp_v4_pre_connect is a call to any bpf hooks, this calls cilium.
* 2. Cilium sees that the address (10.102.48.51) is a service address and rewrites it (in place) with an actual pod IP.
* 3. tcp_v4_connect is called with the rewritten address.

Any rewrite by cilium above directly changes the actual address inside the rpc_xprt struct. So the rpc client now effectively oses track of the service IP. 

When we consequently force a reconnect by using tcpkill, the rpc client will reconnect using the only address it knows about, which is now the pod IP (see the second xs_connect). This means that cilium can no longer rewrite the service address to any new pod IP if that becomes necessary, since it won't see it.

I'm guessing that usually the address here is just a copy of something from userspace and can be rewritten without issue, but since the rpc client is part of the kernel it's able to pass a pointer directly to its internal data instead. This probably breaks an assumption made somewhere.

**Comment by Rid** [comment]
@dhedberg Have you confirmed using bpftrace that on an non-NFS pod, after the tcpkill, `xs_connect` is still using the service IP?

**Comment by dhedberg** [comment]
@Rid xs_connect is an NFS specific function, so you would have to dig for some other workload that uses kernel_connect (might be broken as well) or one that goes through the syscall.

I'm happy to do any further experiments if requested by a developer and of course happy to be corrected if my understanding is incorrect, but at the moment I feel fairly confident to leave it with the above :-)

**Comment by Rid** [comment]
@dhedberg Thanks for the explanation, would you mind sharing the source you used for the bpftrace? I'll test it on the cluster I setup for @brb, it may also help speed up debugging.

**Comment by dhedberg** [comment]
@Rid It's not pretty, but [here are the relevant parts](https://gist.github.com/dhedberg/d17b10dc9522347d860c7ebbb0cc5795).

**Comment by akhilles** [comment]
As a mitigation, would it be reasonable to skip socket LB for Service external IPs? It would be a less efficient datapath as the traffic would have to leave + re-enter the node before it gets load-balanced, but it should work for in-kernel clients.

**Comment by Rid** [comment]
Thanks @brb I was going to reopen this, is there any update regarding https://github.com/cilium/cilium/pull/25169 being merged? It looks like there's a couple of failing tests.

**Comment by jrife** [comment]
> When not using kube-proxy-replacement the mount reconnects automatically and everything works again.

@Rid What configuration did you use with Cilium to get NFS to work? Did you disable `kube-proxy-replacement` by setting it to `disabled` or `partial`? Testing this locally with Cilium `v1.12.6` and can't get NFS to reconnect no matter what I try.

**Comment by Rid** [comment]
@jrife I set `kube-proxy-replacement` to disabled and used `IPVS`, however you need to be mindful of another bug with IPVS: https://github.com/cilium/cilium/issues/18610

In order to overcome that issue we used a script which injects the additional iptables rules and is run on all nodes:

```
#!/bin/bash

CILIUM_HOST_IP=$(ip --json addr show dev cilium_host | jq -r '.[0].addr_info[0].local')

iptables -t nat -A CILIUM_POST_nat -o cilium_host -m mark --mark 0xf00/0xf00 -m comment --comment "hairpin traffic that originated from a local pod" -j SNAT --to-source "${CILIUM_HOST_IP}"
iptables -t nat -D CILIUM_POST_nat -o cilium_host -m mark --mark 0xf00/0xf00 -m conntrack --ctstate DNAT -m comment --comment "hairpin traffic that originated from a local pod" -j SNAT --to-source "${CILIUM_HOST_IP}"

iptables -A CILIUM_OUTPUT -m mark ! --mark 0xe00/0xf00 -m mark ! --mark 0xd00/0xf00 -m mark ! --mark 0xa00/0xe00 -m mark ! --mark 0x800/0xe00 -m mark ! --mark 0xf00/0xf00 -m comment --comment "cilium: host->any mark as from host" -j MARK --set-xmark 0xc00/0xf00
iptables -D CILIUM_OUTPUT -m mark ! --mark 0xe00/0xf00 -m mark ! --mark 0xd00/0xf00 -m mark ! --mark 0xa00/0xe00 -m mark ! --mark 0x800/0xe00 -m comment --comment "cilium: host->any mark as from host" -j MARK --set-xmark 0xc00/0xf00
```

I haven't had time recently, but it might be worth compiling and testing https://github.com/cilium/cilium/pull/25169 to see if it also fixes this issue?

**Comment by jrife** [comment]
@Rid Looks like I'm running `kube-proxy` in iptables mode. I deployed `kube-proxy` and setting `kube-proxy-replacement` to `disabled`, but found that I needed to reboot my nodes before NFS reconnects worked as I wanted them. Before rebooting all my nodes, I tried restarting Cilium agents, restarting my NFS client pod, etc. but nothing worked until I rebooted my nodes. I'm assuming some other state or config was cleared by the reboots but not sure what.

**Comment by jrife** [comment]
[This](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=0bdf399342c5) kernel patch fixes the issue.

**Comment by aditighag** [comment]
> [This](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=0bdf399342c5) kernel patch fixes the issue.

@jrife Nice find! I had a question around the NFS application: does it try to re-connect on an existing socket when remote connectivity breaks? On a related note, cilium intercepts some of other socket calls such as `getpeername` and `bind`, so I wonder if this logic in cilium BPF programs suffices - https://github.com/cilium/cilium/blob/main/bpf/bpf_sock.c#L639?

**Comment by jrife** [comment]
@aditighag Yes, I found that reconnect attempts are made when the remote connectivity breaks.

> I wonder if this logic in cilium BPF programs suffices, or are there cases where your kernel patch would have to be extended for these sockets calls?

I actually drafted an [earlier patch](https://github.com/jrife/linux/commit/a9dc6a333e82d2c4d40855b4637b05a964aa82d5) that performs an address copy for both `kernel_sendmsg` and `kernel_bind`, as in theory they could have the same issue. However,  I ended up not including these in my patch because there weren't any obvious scenarios that I could test that would be broken.

**Comment by borkmann** [comment]
> @aditighag Yes, I found that reconnect attempts are made when the remote connectivity breaks.
> 
> > I wonder if this logic in cilium BPF programs suffices, or are there cases where your kernel patch would have to be extended for these sockets calls?
> 
> I actually drafted an [earlier patch](https://github.com/jrife/linux/commit/a9dc6a333e82d2c4d40855b4637b05a964aa82d5) that performs an address copy for both `kernel_sendmsg` and `kernel_bind`, as in theory they could have the same issue. However, I ended up not including these in my patch because there weren't any obvious scenarios that I could test that would be broken. The challenge with testing these calls is that you can't call them directly from user space, so I'm limited to code paths reachable by using features from user space. For example, testing the change to `kernel_sendmsg` isn't possible with NFS, because the NFS client code doesn't use the `msg_name` field in `msg` even when operating in UDP mode.
> 
> AFAIK the kernel equivalent to `getpeername` isn't at risk, because the address parameter is supposed to be set by the call.

Sounds good, thanks for the clarifications and for taking up this issue! Once your patch lands in mainline I can ping Greg for stable inclusion. I think the one thing left on Cilium side is to document that the restriction is lifted with your patch, I can send a small doc change PR and then we should be able to close the GH issue.

**Comment by jrife** [comment]
@borkmann Looks like the patch was merged [into mainline](https://github.com/torvalds/linux/blob/bd6c11bc43c496cddfc6cf603b5d45365606dbd5/net/socket.c#L3567) today, albeit during the merge window not yet in a release.

**Comment by jrife** [comment]
I discovered this problem exists with SMB mounts as well due to [this](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/fs/smb/client/connect.c#L3044) code:

```C
	rc = socket->ops->connect(socket, saddr, slen,
				  server->noblockcnt ? O_NONBLOCK : 0);
```

which skips `kernel_connect()` completely and just calls `socket->ops->connect()`. I tested a patch that fixes the issue by calling `kernel_connect` instead to make sure `saddr` doesn't get overwritten:

```C
	rc = kernel_connect(socket, saddr, slen, server->noblockcnt ? O_NONBLOCK : 0);
```

I'll send this upstream and work on backporting it. Similar code exists (`socket->ops->connect()`) in the Ceph driver, but I wasn't able to break Ceph yet (testing with Rook).

**Comment by jrife** [comment]
> I discovered this problem exists with SMB mounts as well due to [this](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/fs/smb/client/connect.c#L3044) code:
> 
> ```c
> 	rc = socket->ops->connect(socket, saddr, slen,
> 				  server->noblockcnt ? O_NONBLOCK : 0);
> ```
> 
> which skips `kernel_connect()` and just calls `socket->ops->connect()`. One way to fix it is with this patch:
> 
> ```c
> 	rc = kernel_connect(socket, saddr, slen, server->noblockcnt ? O_NONBLOCK : 0);
> ```
> 
> However, there are a lot of other places in the kernel (Ceph driver included) where a raw call to `socket->ops->connect()` exists, and it's not clear which of these might break when using Cilium. A better solution would be to push the address copy downward a bit in the stack from `kernel_connect` to where `pre_connect()` is called:
> 
> [net/ipv4/af_inet.c](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/net/ipv4/af_inet.c):
> 
> * [`inet_dgram_connect`](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/net/ipv4/af_inet.c#L566)
> * [`inet_stream_connect`](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/net/ipv4/af_inet.c#L734)
> 
> [net/mptcp/protocol.c](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/net/mptcp/protocol.c):
> 
> * [`mptcp_connect`](https://github.com/torvalds/linux/blob/32bf43e4efdb87e0f7e90ba3883e07b8522322ad/net/mptcp/protocol.c#L3603)
> 
> Otherwise it will just turn into a game of whac-a-mole. This does lead to an unnecessary address copy in cases where the code path is reached from `sys_connect()`, but it doesn't seem like this would have an appreciable performance impact.
> 
> FYI @borkmann

[This](https://lore.kernel.org/netdev/20230921234642.1111903-1-jrife@google.com/T/#u) patch series was merged into mainline today which addresses similar issues with the `kernel_bind()`, `kernel_sendmsg()`, and `sock_sendmsg()` in kernel space. I'm following up with several patches to various subsystems such as [Ceph](https://lore.kernel.org/ceph-devel/20231004233827.1274148-1-jrife@google.com/T/#u), and [SMB](https://lore.kernel.org/linux-cifs/20231004011303.979995-1-jrife@google.com/T/#u) so that they use these safe versions of `kernel_bind()` and `kernel_connect()` instead of raw calls to `ops->connect()` and `ops->bind()`.
